proposals/2023-08-21-utf8.md

```diff
@@ -0,0 +1,258 @@
+
+There may also be differences between PromQL and the exposition format as to what characters must be escaped. To reduce confusion we may require escaping for characters that need not be escaped in one language but do need to be escaped in the other.
+
+### Templating syntax
+
+Templating is used in alerts, and for dashboards that want to reference metric values. Go templating already supports quoting by way of the `index` keyword. Whereas labels are usually denoted as `{{ $labels.my_label_name }}`, the quote syntax is `{{ index $labels "my.label.name" }}`. No code changes are required.
+
+## Backwards-compatibility
+
+There a few situations where we need to consider backwards compatibility:
+
+1. Old metrics provider, new backend: Old providers will not serve UTF-8 metric names, so nothing needs to be done here.
+2. New metrics provider, old backend: Old versions of Prometheus will throw an error if they see the new syntax. Therefore we must provide a way for the scraped target to serve previously-invalid metrics in a form that the old database can process.
+3. New backend, old query client: Errors will occur if an old client is fed UTF-8 metric names. Therefore we must provide a way for Prometheus to escape metric names such that old clients can query for and read them.
+4. Old backend, new query client: as with the first case, nothing needs to be done.
+
+### Text escaping
+
+We propose an escaping syntax for metric names that contain UTF-8 characters for systems that do not support UTF-8. This system can be used in queries and responses to provide compatibility between the systems.
+
+* Prefix the string with `U__`.
+* All non-valid characters (i.e. characters other than letters, numbers, and underscores) will be encoded as underscores surrounding the unicode value, like `_1F60A_`.
+* All pre-existing underscores will become doubled: `__`.
+* If a string should start with "U__" already, it will need to be escaped: `U___55_____`. (That's `U__` + `_55_` (for 'U') + `__` + `__`).
+
+#### Mixed-Block Scenario 
+
+We must consider an edge case, in which a newer client persists metrics to disk in an older database that does not support UTF-8. Those metrics will be written to disk with the U__ escaping format. If, later, the user upgrades their database software, new metrics will be written with the native UTF-8 characters. At query time, there will be a problem: newer blocks were written with UTF-8 and older blocks were written with the escaping format. The query code will not know which encoding to look for.
+
+To avoid this confusion we propose to bump the version number in the tsdb meta.json file. On a per-block basis we will then know whether to look for UTF-8 metric names in native encoding (bumped version number 2) or in U__-escaped naming (current version 1).
+
+If a metric is written as "U__" but does not pass the unescaping algorithm, then we assume it was meant to be read as-is and let it pass through.
```

Comment:
I think this is fine now, but it should still call out the remaining edge case. Maybe add something like:

"In turn, this means that an old metric or label name containing a valid escape sequence by chance would be parsed differently after enabling UTF-8 support. For example, a metric called `U___263a_` would appear as `☺` after the upgrade (if read from an old storage chunk, scraped from an old target, or received from an old remote-write client). This is an extremely unlikely scenario in practice so that the theoretically possible implicit name change is acceptable."

Reply:
adopted as-is